pkg/controller/backup_deletion_controller.go

```diff
@@ -0,0 +1,290 @@
+		log.Debug("Unable to find DeleteBackupRequest")
+		return nil
+	}
+	if err != nil {
+		return errors.Wrap(err, "error getting DeleteBackupRequest")
+	}
+
+	switch req.Status.Phase {
+	case "", v1.DeleteBackupRequestPhaseNew:
+		// Don't mutate the shared cache
+		reqCopy := req.DeepCopy()
+		return c.processRequest(reqCopy)
+	}
+
+	return nil
+}
+
+func (c *backupDeletionController) processRequest(req *v1.DeleteBackupRequest) error {
+	log := c.logger.WithFields(logrus.Fields{
+		"namespace": req.Namespace,
+		"name":      req.Spec.BackupName,
+	})
+
+	var err error
+
+	// Update status to InProgress
+	req, err = c.patchDeleteBackupRequest(req, func(r *v1.DeleteBackupRequest) {
+		r.Status.Phase = v1.DeleteBackupRequestPhaseInProgress
+	})
+	if err != nil {
+		return err
+	}
+
+	// Get the backup we're trying to delete
+	backup, err := c.backupClient.Backups(req.Namespace).Get(req.Spec.BackupName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// Couldn't find backup - update status to Processed and record the not-found error
+		req, err = c.patchDeleteBackupRequest(req, func(r *v1.DeleteBackupRequest) {
+			r.Status.Phase = v1.DeleteBackupRequestPhaseProcessed
+			r.Status.Errors = []string{"backup not found"}
+		})
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+	if err != nil {
+		return errors.Wrap(err, "error getting Backup")
+	}
+
+	// If the backup includes snapshots but we don't currently have a PVProvider, we don't
+	// want to orphan the snapshots so skip deletion.
+	if c.snapshotService == nil && len(backup.Status.VolumeBackups) > 0 {
+		req, err = c.patchDeleteBackupRequest(req, func(r *v1.DeleteBackupRequest) {
+			r.Status.Phase = v1.DeleteBackupRequestPhaseProcessed
+			r.Status.Errors = []string{"unable to delete backup because it includes PV snapshots and Ark is not configured with a PersistentVolumeProvider"}
+		})
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// Set backup status to Deleting
+	backup, err = c.patchBackup(backup, func(b *v1.Backup) {
+		b.Status.Phase = v1.BackupPhaseDeleting
+	})
+	if err != nil {
+		log.WithError(errors.WithStack(err)).Error("Error setting backup phase to deleting")
+	}
+
+	var errs []string
+
+	// Try to delete snapshots
+	log.Info("Removing PV snapshots")
+	for _, volumeBackup := range backup.Status.VolumeBackups {
+		log.WithField("snapshotID", volumeBackup.SnapshotID).Info("Removing snapshot associated with backup")
+		if err := c.snapshotService.DeleteSnapshot(volumeBackup.SnapshotID); err != nil {
+			errs = append(errs, errors.Wrapf(err, "error deleting snapshot %s", volumeBackup.SnapshotID).Error())
+		}
+	}
+
+	// Try to delete backup from object storage
+	log.Info("Removing backup from object storage")
+	if err := c.backupService.DeleteBackupDir(c.bucket, backup.Name); err != nil {
+		errs = append(errs, errors.Wrap(err, "error deleting backup from object storage").Error())
+	}
+
+	// Try to delete restores
+	log.Info("Removing restores")
```

Comment:
IMO we should remove restores when explicitly deleting a backup, but I have seen others mention that retaining restores as a matter of record would be beneficial. The restores would then be cleaned up via the GC controller. I think that's also a reasonable approach, wdyt?

I don't think this should be an option, though, we should only do one or the other.

Reply:
I'm seeing now that the GC controller uses the `BackupDeleteRequest`s to do a delete, so I suppose the 2nd approach would require shuffling some logic around. I'm fine with it as is.